src/peer_manager.rs

```diff
@@ -971,23 +1036,30 @@ impl PeerManager {
         let _ = self.peers.insert(peer.pub_id, peer);
     }
 
-    /// Removes the given peer. Returns whether the peer was actually present.
-    pub fn remove_peer(&mut self, pub_id: &PublicId) -> bool {
-        let remove_candidate = match self.candidate {
-            Candidate::None => false,
-            Candidate::AcceptedForResourceProof { ref old_pub_id, .. } => {
-                // only consider candidate cleanup via old_id if candidate is also expired.
-                // else candidate may simply be restarting.
-                old_pub_id == pub_id && self.candidate.is_expired()
+    /// Forget about the current candidate.
+    pub fn reset_candidate(&mut self) {
+        self.candidate = Candidate::None;
+    }
+
+    /// Forget about the current candidate if it is a member of the given section.
+    pub fn reset_candidate_member_of(&mut self, members: &BTreeSet<PublicId>) {
+        if let Candidate::AcceptedWaitingSectionInfo { ref new_pub_id } = self.candidate {
+            if members.contains(new_pub_id) {
+                self.candidate = Candidate::None;
             }
-            Candidate::ResourceProof { new_pub_id, .. } => new_pub_id == *pub_id,
-        };
+        }
+    }
 
-        if remove_candidate {
+    /// Forget about the current candidate if it matches the purged old public id.
+    pub fn reset_candidate_with_old_public_id(&mut self, old_pub_id: &PublicId) {
+        if Some(old_pub_id) == self.candidate.old_pub_id() {
             self.candidate = Candidate::None;
         }
+    }
 
-        self.peers.remove(pub_id).is_some() || remove_candidate
+    /// Removes the given peer. Returns whether the peer was actually present.
+    pub fn remove_peer(&mut self, pub_id: &PublicId) -> bool {
+        self.peers.remove(pub_id).is_some()
```

Comment:
What happens now when we call `remove_peer` with a `pub_id` of the candidate? Aren't we going to end up in inconsistent state, because `self.candidate` will be still set, but the corresponding entry in `self.peers` will be gone? Maybe we shouldn't remove the entry from `self.peers` in this case?

Reply:
> Aren't we going to end up in inconsistent state, because self.candidate will be still set, but the corresponding entry in self.peers will be gone?
that should be an acceptable scenario in general as in the section has a candidate that maybe some members cant connect to and thus dont have in peer_mgr. so the candidate info should be uncoupled to the candidates state in peer_mgr itself in terms of connected or not. So while I'm not sure if we call remove_peer for a valid candidate, even if we did we should expect it to only result in an eventual purge candidate for that peer later and them not being accepted into the section.